app/src/main/java/io/apicurio/registry/store/RegistryStorage.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2019 Red Hat
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.apicurio.registry.store;
+
+import java.util.List;
+
+import io.apicurio.registry.rest.beans.ArtifactMetaData;
+import io.apicurio.registry.rest.beans.EditableMetaData;
+import io.apicurio.registry.rest.beans.Rule;
+import io.apicurio.registry.rest.beans.VersionMetaData;
+
+/**
+ * The storage layer for the registry.
+ * @author eric.wittmann@gmail.com
+ */
+public interface RegistryStorage {
+
+    public ArtifactMetaData createArtifact(String artifactId, String content) throws ArtifactAlreadyExistsException, RegistryStorageException;
```

Comment:
What about if we used pure byte[] for the content?
I remember having issues once with encoding ... 
But I guess if we used UTF-8 all the time or as default.

Reply:
We need a second method that generates `artifactId`, not expecting one from the user, for `POST` to `/artifacts`. I think it's most appropriate to generate the IDs at the persistence level.

Reply:
This could be done if artifactId param is null / not passed.

Reply:
ArtifactMetaData should always return artifactId -- even if user-passed.

Reply:
Good points here - there are actually multiple issues that you guys have raised.  One is how does the user specify a String artifact ID (in the REST API) if they want to name it.  A custom request header, for example.

I agree that generating the ID in the storage layer is correct.  I also agree that passing `null` for the `artifactId` parameter (for this method) could indicate that no client-provided ID was sent.  This implies that client-provided IDs are **optional**.  I'm OK with that if you guys are.

Reply:
I guess the problem here is if the artifactId comes from path parameter, not from request parameter.
Is that a problem for our REST endpoints?

Reply:
One more note - It's just a habit I suppose, but I'd use DTOs more, i.e. not call the `createArtifact` method with the content directly. But that depends on how much layering in the application architecture each developer prefers:)

Reply:
@alesj Right - the issue is on `POST` `/artifacts` - how to pass the client-provided Artifact ID into that endpoint.  Options are:

* Query parameter
* Request header
* Change the request body to be something more confluent-like (a JSON payload with the content as a property of that JSON structure)

I don't like option 3.  So my suggestion is that we make the client-provided ID **optional** but allow it to be passed in via either a query param or a header param (or perhaps support both options).

Does that make sense?

Reply:
@jsenko I'm not against using DTOs more.

Reply:
@EricWittmann yeah -- "my suggestion is that we make the client-provided ID optional but allow it to be passed in via either a query param or a header param (or perhaps support both options)" -- sounds OK

@jsenko since this is internal API / SPI, we could change that (simple parameter to DTO) later -- "premature optimization is root of all evil ... " :-)